### ISSUE TITLE
ENH: memory-efficient `np.cov` for `n_samples >> n_features`

### ISSUE THREAD

**Issue by bruAristimunha** [numpy/numpy#31292]
### Proposed new feature or change:

`np.cov` centers its input by allocating `m - mean(m)`, a copy the size of `m`. When observations vastly outnumber variables, that copy dominates peak memory and wall time; the covariance matrix itself is tiny.

I and potentially scikit-learn (cc. @ogrisel) would like to add an opt-in keyword that uses the algebraically equivalent post-hoc formula:

    cov = (m @ m.conj().T) / fact - (N / fact) * avg @ avg.conj().T

It is already implemented here on scikit-learn: 

https://github.com/scikit-learn/scikit-learn/blob/94af1dff678e8a7a6fc627b85a5489980c063331/sklearn/decomposition/_pca.py#L604-L610

I think this optimization should be put here, and we have many historical examples of this type of optimization on numpy. From the historical perspective:

1. In the PR #6403 (merged 2015), already halved `np.cov` memory; the ' m-mean` copy is what's left. 
2. #13199 tracks the same class of problem for `np.var` and `np.std` (open). 
3. #6231 proposes Welford's algorithm; that's orthogonal, because post-hoc pairs with BLAS `gemm`/`syrk` and Welford does not.

#1696, #9631, #4694 cover the stability/memory tradeoff for variance-like reductions.

Scikit-learn reimplements this locally and already shows strong benchmark results with this. `PCA(svd_solver="covariance_eigh")` does `C = X.T @ X - N * mean mean.T; C /= (N - 1)` ([source][sklearn-pca]) to avoid the copy. [scikit-learn#33573][] ([review comment][grisel-comment]) asks for this to move upstream.

The small trade-off is that the post-hoc formula loses precision when `|avg|^2 / var(m)` is large; the reference that discusses this is Chan, T. F., Golub, G. H., & LeVeque, R. J. (1983).

So the default stays safe. Callers who know their data is well-conditioned opt in.

## Proposal

    np.cov(m, ..., *, method="pre-center")   # or "post-hoc"

Weighted (`fweights`, `aweights`) and complex cases generalize without structural change. `corrcoef` could follow, but not in the first PR.

Chan, T. F., Golub, G. H., & LeVeque, R. J. (1983). Algorithms for computing the sample variance: Analysis and recommendations. The American Statistician, 37(3), 242-247.

[#6403]: https://github.com/numpy/numpy/pull/6403
[#13199]: https://github.com/numpy/numpy/issues/13199
[#6231]: https://github.com/numpy/numpy/issues/6231
[#1696]: https://github.com/numpy/numpy/issues/1696
[#9631]: https://github.com/numpy/numpy/issues/9631
[#4694]: https://github.com/numpy/numpy/issues/4694
[sklearn-pca]: https://github.com/scikit-learn/scikit-learn/blob/94af1dff678e8a7a6fc627b85a5489980c063331/sklearn/decomposition/_pca.py#L604-L610
[scikit-learn#33573]: https://github.com/scikit-learn/scikit-learn/pull/33573
[grisel-comment]: https://github.com/scikit-learn/scikit-learn/pull/33573#discussion_r3101245545
[pytorch/ignite#3662]: https://github.com/pytorch/ignite/issues/3662

cc. @thomasjpfan, @ogrisel, @qbarthelemy, @betatim, @rgommers, @lucascolley.

and FYI: @seberg and @eric-wieser

**Comment by bruAristimunha** [comment]
hey @seberg, @charris 

I was wondering, would you happen to have any input? I sent the issue to the numpy list email, but I didn't get any input. 

It is my first time contributing here, so if there are any specific meetings, procedures, or discussions I may need to attend, please let me know.

Deeply appreciate your time and patiente 🙏🏽

**Comment by seberg** [comment]
Sorry, I haven't thought about it much.  I don't particularly love adding a new `method=` but a strong user like scikit-learn might be a good enough reason; it's just not something we have often.  So I guess the question is how often this trade-off is OK, I remember pandas used to do that and was happy to be faster than NumPy, but eventually stopped, because it is imprecise often.

**Comment by bruAristimunha** [comment]
Hey @seberg!

I think it's worth it, and I can back that up with numbers (happy to share a benchmark notebook). On a `(16, 5_000_000)` float64 input:

|             | time   | peak alloc |
| ----------- | -----: | ---------: |
| `np.cov(m)` | 136 ms | 640 MB     |
| post-hoc    |  69 ms | ~0 MB      |

@ogrisel had run many, many tests on scikit-learn, and this small trick unlocks the PCA for the largest dataset. From what I saw, it is also a parameter in scikit-learn's PCA.

The maximum absolute deviation from the current implementation is ~2e-14. Beyond the speedup, this opens up matrix sizes that don't fit today: dense covariance or large dataset, where the `n_samples × n_features` copy is the blocker, not the matmul.

This changes nothing for users who don't opt in. The default stays the current centered two-pass. The new behavior only activates if the caller explicitly passes `method=...`. I think it is a nice tradeoff and will not affect the user et all.

**Comment by bruAristimunha** [comment]
The diff would be super small too!

I was playing with some numbers, and this is what I changed in te file: numpy/lib/_function_base_impl.py
```diff
@@ signature
-def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
-        aweights=None, *, dtype=None):
+def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
+        aweights=None, *, dtype=None, method="pre-center"):

@@ safe-copy site
-    X = array(m, ndmin=2, dtype=dtype)
+    X = array(m, ndmin=2, dtype=dtype,
+              copy=True if method == "pre-center" else None)

@@ compute site
-    X -= avg[:, None]
-    X_T = X.T if w is None else (X * w).T
-    c = dot(X, X_T.conj())
+    if method == "pre-center":
+        X -= avg[:, None]
+        X_T = X.T if w is None else (X * w).T
+        c = dot(X, X_T.conj())
+    else:
+        # X @ (X*w)^T - w_sum * outer(mu, mu)
+        #   = sum_i w_i (x_i - mu)(x_i - mu)^T
+        X_T = X.T if w is None else (X * w).T
+        c = dot(X, X_T.conj())
+        c -= w_sum * np.outer(avg, avg.conj())
     c *= np.true_divide(1, fact)
```

**Comment by bruAristimunha** [comment]
About the panda, many thanks for the point. I tried to dig with git blame to see the history. Looks likes was like this: 

1. 2012 (@wesm, original `algos.pyx::nancorr`): two-pass centered. Compute means in pass 1, accumulate `(x − μ)(y − μ)` in pass 2.
2. 2020 ([#37453](https://github.com/pandas-dev/pandas/pull/37453), @phofl): switched to Welford one-pass for numerical stability
3. 2021 ([#42761](https://github.com/pandas-dev/pandas/pull/42761), @mzeitlin11): PERF, in the no-NaN fast path, precompute Welford means and ssqds once per column, then a classical centered pass per pair for the cross term `(vx − means[xi]) * (vy − means[yi])`. The NaN path stayed on Welford. So this was a precomputed-means + centered-cross optimization, not a switch to the naïve form.
4. 2022 ([#45646](https://github.com/pandas-dev/pandas/pull/45646), @phofl): reverted #42761 after [issue #45640](https://github.com/pandas-dev/pandas/issues/45640). On `rng.normal(size=(100, 5))`, `df.corr()` produced a diagonal of `1 + 4.4e-16` instead of exactly `1.0`, which broke `scipy.spatial.distance.squareform(1 - df.corr().abs())`. 

and looks like, nowadays on pandas's current code: `DataFrame.cov` delegates to `np.cov` whenever the input has no NaNs (`pandas/core/frame.py:15434`). Their Cython Welford kernel only runs when masking is required. So every clean-data `df.cov()` call pays the same memory cost we're discussing here. 

From what I read, please correct me if I made any bad interpretation of the team from pandas, the post-hoc trick was not used there `(M @ Mᵀ)/N − outer(μ, μ)` (aka what I am calling "post-hoc" in the issue). In pandas, both algorithms they shipped (two-pass centered, then Welford) and the reverted optimization use centered subtraction. 

What do you think, @seberg? And can you input your experience @ogrisel?

**Comment by betatim** [comment]
Did you run your experiments regarding the precision using `float32` as well? I did a few experiments (and asked AI to do some) and it feels like with `float32` the problem of precision is more serious.

In an ideal universe we'd find a way to not introduce a new argument for `np.cov()`. I'm wondering if there is a way to figure out cases for which we could always use the post-hoc centering method. So far I have no ideas :(

**Comment by bruAristimunha** [comment]
Hard question @betatim!

I conducted a deeper study of precision error using a post hoc approach.

Speed and memory gains are large across the board.

<img width="2323" height="903" alt="Image" src="https://github.com/user-attachments/assets/33a805ef-6223-4654-8450-1fae9a4a0c5e" />

And on well-conditioned data, precision stays at machine ε, indistinguishable from the default.

<img width="1523" height="883" alt="Image" src="https://github.com/user-attachments/assets/96cced4e-e838-4e25-a9c1-d2d2525ab9ff" />

Going to the edge case: the relative error grows quadratically with μ/σ (i.e., a straight line of slope 2 on the log-log plot), matching the Chan-Golub-LeVeque κ² ≈ 1 + (μ/σ)² bound. Only large means with small variances are a problem.

- float32: safe to μ/σ ≈ 10. Loses around μ/σ ≈ 1000.
- float64: safe to μ/σ ≈ 1e4. Crosses 1% relative error around μ/σ ≈ 1e6.

<img width="1523" height="883" alt="Image" src="https://github.com/user-attachments/assets/5835c7c6-bfec-47fa-9adf-46861af09dea" />

To mitigate this, the `metod="post-hoc"` docstring would carry the warning. Something like:

> *Loses precision when ‖μ‖²/Var(X) is large; on float32 this becomes material around μ/σ ≈ 10, on float64 around μ/σ ≈ 10⁵. Use on standardized or already mean-centered data.*

Answer your question: yes, we *can* auto-detect. Compute `max((μ/σ)²) · ε` in one passoverm'  and dispatch on the threshold. The catch is that this check costs roughly half what centering already costs, so a fully automatic dispatch effectively gives back the speedup. 

My honest answer is `method=` opt-in with the docstring warning, and possibly expose the diagnostic as a helper for callers who want to probe their data cheaply before deciding.

Thinking about the API perspective: similar opt-in surface already exists in numpy. `np.linalg.pinv(hermitian=True)` ([#12693](https://github.com/numpy/numpy/pull/12693), pushed by @eric-wieser) is the fast-path-with-precondition pattern. `np.quantile(method=...)` ([#19857](https://github.com/numpy/numpy/pull/19857), Abel Aoun, merged by @seberg) is a 9-option `method=` string. `np.linalg.lstsq(rcond=...)` exposes a numerical-conditioning knob, i feel this is the most similar parallel.

Happy to put up a draft PR with the gating, the docstring, and the regression tests. @seberg, does that surface area sound reasonable to evaluate?

**Comment by bruAristimunha** [comment]
The zip with the code using the numpy asv for the benchmark and plot function is here: 

[cov-31292.zip](https://github.com/user-attachments/files/27962894/cov-31292.zip)

**Comment by bruAristimunha** [comment]
Hey @seberg!

First of all, many thanks for your time here! It is deeply appreciated!

Could you take a look at the request benchmark?

If necessary, I am happy to join any meeting to further the discussion, as I recently discovered the numpy meeting agenda. 

I would need some micro-assistance to point out which of the meetings I should join: new collaboration, triage meeting, optimization group, or community.

I am happy to continue on GitHub, also.
